[#180192789] Rewrite README (#2796)
alphagov/paas-cf: README.md

# paas-cf

⚠️

> When merging pull requests, please use the [gds-cli](https://github.com/alphagov/gds-cli): `gds git merge-sign alphagov/paas-cf PR_NUMBER`

Reviewer: ```suggestion
When merging pull requests, use the [gds-cli](https://github.com/alphagov/gds-cli): `gds git merge-sign alphagov/paas-cf PR_NUMBER`
```

⚠️

> GOV.UK Platform as a Service (PaaS) CF builds upon the foundations laid out in [`paas-bootstrap`](https://github.com/alphagov/paas-bootstrap) to create a working deployment of [Cloud Foundry](https://www.cloudfoundry.org/) (CF) on VMs for GOV.UK PaaS. It handles the following non-exhaustive list of duties:

Author: Yes I like this better

+ [Deploying CF using Concourse](https://github.com/alphagov/paas-cf/blob/main/concourse/pipelines/create-cloudfoundry.yml)
+ [Configuring CF](https://github.com/alphagov/paas-cf/tree/main/manifests/cf-manifest) based on [`cf-deployment`](https://github.com/cloudfoundry/cf-deployment)
+ [Provisioning AWS resources using Terraform](https://github.com/alphagov/paas-cf/tree/main/terraform/cloudfoundry). This includes

>    + Load balancers

Reviewer: ```suggestion
   + load balancers
```

>    + Databases

Reviewer: ```suggestion
   + databases
```

   + IAM roles and policies
   + DNS records

>    + Networking

Reviewer: ```suggestion
   + networking
```

### Pre-requisites
Before you can get a Cloud Foundry deployment up and running, you will need the following available

+ [ ] A running [`deployer-concourse` instance from `paas-bootstrap`](https://github.com/alphagov/paas-bootstrap)
+ [ ] Make
+ [ ] Ruby >= 2.7
+ [ ] [GDS CLI](https://github.com/alphagov/gds-cli)
+ [ ] Access to `paas-credentials` (private repository) and tools installed
+ [ ] Connection to GDS VPN
+ [ ] Permission to assume the `Admin` role of the relevant AWS account (dev, ci, staging, production)
+ [ ] `AWS_DEFAULT_REGION` environment set the desired region for the environment
+ [ ] [cf CLI](https://docs.cloudfoundry.org/cf-cli/install-go-cli.html) >=7

### Deploy Cloud Foundry

> These instructions contain placeholders where the exact command may vary. The below table explains the purpose of those placeholders

Reviewer: ```suggestion
These instructions contain placeholders where the exact command may vary. The below table explains the purpose of those placeholders:
```

| Placeholder   | Purpose                                                                                                                                                                                                           |
| ------------- | ------------------------------------------|
| `$ACCOUNT`    | The AWS account being targeted (for example, `dev`, `staging`)|
| `$ENV` | The name of the environment being targeted. In the case of short lived development environments, this should have a value of `dev`, and the specific environment is set by the `DEPLOY_ENV` environment variable (max 8 chars)|

> 1. Log in to [Credhub](https://docs.cloudfoundry.org/credhub/) in the environment by running this and following the instructions on screen. This will take you in to a new shell session.

Reviewer: ```suggestion
1. Log in to [CredHub](https://docs.cloudfoundry.org/credhub/) in the environment by running this and following the instructions on screen. This will take you into a new shell session.
```

   ```shell
   gds aws paas-$ACCOUNT-admin -- make $ENV credhub
   ```

> 1. Upload the secrets to Credhub from the Credhub shell session.

Reviewer: ```suggestion
1. Upload the secrets to CredHub from the CredHub shell session.
```

   ```shell
   make $ENV upload-all-secrets
   ```

>    Note: you do not need to use GDS CLI here because the Credhub shell session contains the AWS credentials in environment variables

Reviewer: ```suggestion
   Note: you do not need to use GDS CLI here because the CredHub shell session contains the AWS credentials in environment variables
```

> 1. Exit the Credhub shell session

Reviewer: ```suggestion
1. Exit the CredHub shell session
```

> 1. Deploy the pipeline configurations using `make`. This will upload/update the pipelines. Select the target based on which AWS account you want to work with:

Reviewer: ```suggestion
1. Deploy the pipeline configurations using `make`. This will upload or update the pipelines. Select the target based on which AWS account you want to work with:
```

   ```shell
   gds aws paas-$ACCOUNT-admin -- make $ENV pipelines
   ```
1. Log in to Concourse. See the [Accessing Concourse](#accessing-concourse).

1. Tun the `generate-paas-admin-git-keys`, `generate-paas-billing-git-keys` and `generate-paas-aiven-broker-git-keys` jobs in the job group `operator`. This will generate and store some SSH keys needed by other jobs.

1. Run the `create-cloudfoundry` pipeline, starting from the left-hand `pipeline-lock` job. This will configure and deploy Cloud Foundry. It might take a couple of hours to complete.

## Cloud Foundry deployment configuration options

> There are a handful of configuration options which can change a Cloud Foundry deployment which can only be set at pipeline level. Each of the properties in the below table should be set as Make variables when setting the pipelines

Reviewer: Should we have the backticks around 'make' here?

Author: I'm not sure what the rule are here. Make is the name of the product, and `make` is the name of the program.

Reviewer: It should be OK as it is, then. Perhaps we could link to the product's docs here?

Author: Good idea!

> There are a handful of configuration options which can change a Cloud Foundry deployment which can only be set at pipeline level. Each of the properties in the below table should be set as Make variables when setting the pipelines

Reviewer: ```suggestion
There are a handful of configuration options which can change a Cloud Foundry deployment which can only be set at pipeline level. Each of the properties in the below table should be set as Make variables when setting the pipelines:
```

```
gds aws paas-$ACCOUNT-admin -- make $ENV pipelines VAR=value
```

| Property (VAR) | Type | Default | Description |
| -- | -- | -- | -- |
| `BRANCH` | String | `main` | Sets the `paas-cf` branch which will be used in the pipeline |
| `DEPLOY_ENV` | String | `null` for short-lived dev envs, fixed in `Makefile` for other envs | Sets the name of the environment |
| `SELF_UPDATE_PIPELINE` | Bool | `true` | Whether the pipeline should update its own definition from the current branch at runtime. Disable this if you're making a pipeline change which has not been pushed to branch yet |
| `SLIM_DEV_DEPLOYMET` | Bool| `true` in dev, `false` elsewhere | If `true`, reduces the number and size of VMs created for each component to 2. In dev, set this to `false` when testing the impact of a change on platform availability |
|`DISABLED_AZS` | String list, space separated | `""` | <p>Disables the given availability zones in Bosh. This is used when an availability zone goes away, and we need to redistribute virtual machines away from that AZ. </p><p> Set to a value like `"z1 z2"`</p>|

> |`ENABLE_AUTODELETE` | Bool | `true` in dev, `false` elsewhere | <p>If `true`, deploys a pipeline which tears down Cloud Foundry at 8pm each day as a cost saving measure.</p><p>This should absolutely **never** be set to `true` in a staging or production deployment</p> |

Reviewer: ```suggestion
|`ENABLE_AUTODELETE` | Bool | `true` in dev, `false` elsewhere | <p>If `true`, deploys a pipeline which tears down Cloud Foundry at 8pm each day as a cost saving measure.</p><p>This should absolutely never be set to `true` in a staging or production deployment</p> |
```

> |`ENABLE_DESTROY` | Bool | `true` in dev, `false` elsewhere |<p>If `true`, deploys a pipeline which, when run, will completely destroy Cloud Foundry and all of it's data</p><p>This should absolutely **never** be set to `true` in a staging or production deployment</p>|

Reviewer: ```suggestion
|`ENABLE_DESTROY` | Bool | `true` in dev, `false` elsewhere |<p>If `true`, deploys a pipeline which, when run, will completely destroy Cloud Foundry and all of its data</p><p>This should absolutely never be set to `true` in a staging or production deployment</p>|
```

| Directory | Purpose | I will need this when .. |
| -- | -- | -- |
| `concourse/pipelines/` | YAML definitions of the Concourse pipelines | I want to make a change to how the platform is deployed, monitored, or torn down |
| `config/billing/` | The scripts and static files used to generate configuration for the billing system.| <ul><li>I'm adding a new backing service, so that I can set how much it costs.</li><li>The VAT rate has changed </li><li>The cost of an AWS resource has changed</li></ul>|
| `manifests/cf-manifest/` | The Bosh manifest configuration for Cloud Foundry | See specific directories below |
| `manifests/cf-manifest/operations.d/` | Customisations applied to `cf-deployment`, applicable to all environments | <ul><li>I want to make a configuration change that will affect every environment</li><li>I want to deploy a new piece of software with a Bosh release</li></ul>
| `manifests/cf-manifest/operations` | Customisations applied to `cf-deployment` [based on some condition](https://github.com/alphagov/paas-cf/blob/main/manifests/cf-manifest/scripts/generate-manifest.sh#L18) | I want to make a configuration change that will only be applied in certain circumstances |

> |`manifests/cf-manifest/spec`| Unit tests applied to the generated manifest file | I want to ensure a property of the manifest is not invalidated (for example, correct number of instances of some VM) |

Reviewer: ```suggestion
|`manifests/cf-manifest/spec`| Unit tests applied to the generated manifest file | I want to make sure a property of the manifest is not invalidated (for example, correct number of instances of some VM) |
```

> |`manifests/cf-manifest/spec`| Unit tests applied to the generated manifest file | I want to ensure a property of the manifest is not invalidated (for example, correct number of instances of some VM) |

Reviewer: Should VM be plural here?

Author: Hmmm. Good question. I was trying to talk about ensuring we have the right number of instances of a component, but without getting in to all the nuance of being able to have more than 1 component on a machine. Happy to take guidance.

Author: Quick poll with the team suggests they prefer "some VM" over "some VMs" or anything else
